### ISSUE TITLE
Redux-form pass onSubmit

### ISSUE THREAD

**Issue by waynebrantley** [DefinitelyTyped/DefinitelyTyped#15026]
@CarsonF

Take a look at this accepted answer.
http://stackoverflow.com/questions/33237818/react-redux-whats-the-best-way-to-handle-crud-in-a-form-component

I tried to do this using your definitions, but there is no way to send the onSubmit:
```
  render() {
    return (
      <div id="app">
        <h1>App</h1>
        <ContactForm onSubmit={this.handleSubmit.bind(this)}/>
      </div>
    );
  }
```

**Comment by CarsonF** [comment]
I was notified about this from your mention just fine. I've just been busy lately. 

Yes the definition expects the form to be completely self contained. What are you trying to do?

The example clears the form on submit. Here's how you can do that inside the form:
```ts
class ContactForm extends React.Component {
  render() {
    const { fields: {name, address, phone}, handleSubmit } = this.props;
    return (
      <form onSubmit={handleSubmit}>
        <label>Name</label>
        <input type="text" {...name}/>
        {name.error && name.touched && <div>{name.error}</div>}

        <label>Address</label>
        <input type="text" {...address} />
        {address.error && address.touched && <div>{address.error}</div>}

        <label>Phone</label>
        <input type="text" {...phone}/>
        {phone.error && phone.touched && <div>{phone.error}</div>}

        <button onClick={handleSubmit}>Submit</button>
      </form>
    );
  }
}

ContactForm = reduxForm({
  form: 'contact',                      // the name of your form and the key to
                                        // where your form's state will be mounted
  fields: ['name', 'address', 'phone'], // a list of all your fields in your form

  onSubmit: (values, dispatch, props) => {
    props.initialize({}); // clear form
  },

  validate: validateContact             // a synchronous validation function
})(ContactForm);

export default ContactForm;
```

**Comment by waynebrantley** [comment]
Bottom line is I am trying to NOT have the submit code part of the form.  Let the form be a component used by other components and they will provide the submit.   However, it seems that onSubmit is not properly provided

```js
 interface IAddress {
    city: string;
    state: string;
    zip: string
}
export class AddAddressForm extends Component<FormProps<IAddress, void, void>, void> {
    FormToDisplay: any = () => {
        const { handleSubmit } = this.props;
        return (
            <form onSubmit={handleSubmit}>
                      .....blah......
                    <button onClick={handleSubmit}>Submit</button>
            </form>
        );
    }
    render() {
        return (
            <div className="panel panel-default">
                <div className="panel-heading">
                    <h2 className="panel-title"> address</h2>
                </div>
                <div className="panel-body">
                    <div className="form">
                        {this.FormToDisplay()}
                    </div>
                </div>
            </div>
        );
    }
}
```

Now, in a another page, we can import that AddAddressForm component.
In this form, I want to somehow use that as a component...where the onsubmit is hooked up to a local method on the ``AddAddress`` component:
```js
class AddAddress extends Component<{}, {}> {
  AddAddressFormReduxedLocally:any  = reduxForm({ form: 'myform', 
                                                  onSubmit: (values, dispatch, props) => {
                                                        props.initialize({}); // clear form
                                                }})(AddAddressForm);
  render() {
    return (
        this.AddAddressFormReduxedLocally
    );
  }
}

```
When you run this, you get "Uncaught Error: You must either pass handleSubmit() an onSubmit function or pass onSubmit as a prop"

**Comment by waynebrantley** [comment]
See how in these docs:
http://redux-form.com/6.5.0/docs/GettingStarted.md/

They can create the reduxForm and not specify the onSubmit.  Then later, they just use it as a normal component.  If I do that with your definitions, I cannot specify an onSubmit from what ``reduxForm(`` returns?

**Comment by CarsonF** [comment]
Hmm I see. Well for now you could just add `onSubmit` to your `AddAddressForm` props.

**Comment by waynebrantley** [comment]
Is this as easy as adding ``HTMLFormElement`` to this interface
```
export interface FormDecorator<FormData extends DataShape, P, S> {
    <T extends (typeof Component | StatelessComponent<any>)>(component: T): 
        T & Form<FormData, P, S>;
}
```
so would end up as:
```
export interface FormDecorator<FormData extends DataShape, P, S> {
    <T extends (typeof Component | StatelessComponent<any>)>(component: T): 
        T & Form<FormData, P, S> & HTMLFormElement;
}
```
Or is there a bunch more to it?

**Comment by CarsonF** [comment]
No changes to the redux form definitions are needed.

```ts
interface IAddress {
    city: string;
    state: string;
    zip: string
}

interface AddAddressFormProps {
  onSubmit?: SubmitHandler<IAddress, AddAddressFormProps, void>;
}

export class AddAddressForm extends Component<AddAddressFormProps, void> {
    FormToDisplay: any = () => {
        const { handleSubmit } = this.props;
        return (
            <form onSubmit={handleSubmit}>
                      .....blah......
                    <button onClick={handleSubmit}>Submit</button>
            </form>
        );
    }
    render() {
        return (
            <div className="panel panel-default">
                <div className="panel-heading">
                    <h2 className="panel-title"> address</h2>
                </div>
                <div className="panel-body">
                    <div className="form">
                        {this.FormToDisplay()}
                    </div>
                </div>
            </div>
        );
    }
}
```

But I still have to say this feels like the wrong approach. The forms are meant to be used uniquely. If you're just looking for an to abstract common fields you should take a look at the [FormSection](http://redux-form.com/6.5.0/docs/api/FormSection.md/) component.

**Comment by waynebrantley** [comment]
Not really looking for an abstract.  The forms are really meant to be stateless.  The submit function sometimes needs to be in another stateful component that can act on the submit and do different things.

I don't think it is the wrong approach being that is how the 'getting started' works and most examples I find.  They define the form as a stateless component....then use that component giving it a submit.

I however am no expert?  Just going off the http://redux-form.com/6.5.0/docs/GettingStarted.md/ and things like the initial stack overflow link I gave you.

The bottom line is the ReduxForm() call simply decorates a form, so you still should end up with a form as far as typescript is concerned?   That seems to be the right approach?

(Ultimately using typescript has made using ReactForm 10000x more difficult than just straight JSX....and I really want to use and like typescript)

**Comment by waynebrantley** [comment]
Re-reading your last comment, still not sure we are completely in sync. 
I should be able to do this:
Define a stateless Form component based on ReduxForm, just like in the getting started:
(And with your definitions, I can do this)
contactForm.js
```js
interface IAddress {
    city: string;
    state: string;
    zip: string
}

class ContactForm extends Component<FormProps<IAddress, void, void>, void> {
  render() {
    const { error, pristine, reset, submitting, handleSubmit } = this.props;
    return (
      <form onSubmit={handleSubmit}>
        <label>Name</label>
        <input type="text" {...name}/>
        {name.error && name.touched && <div>{name.error}</div>}

        <label>Address</label>
        <input type="text" {...address} />
        {address.error && address.touched && <div>{address.error}</div>}

        <label>Phone</label>
        <input type="text" {...phone}/>
        {phone.error && phone.touched && <div>{phone.error}</div>}

        <button onClick={handleSubmit}>Submit</button>
      </form>
    );
  }
}

export const default reduxForm({ form: 'contactForm' })(ContactForm);
```

Now, in a second file I should be able to use that stateless form:
```js
import * as ContactForm from 'contactForm.js';

class AddAddress extends Component<{}, {}> {
  render() {
     let handleSubmit = (data) => {
          console.warn('submitted!');
          const updated = this.state.data.slice(0);
          updated.push(data);
         this.setState({ data: data});
    }
    return (
        <ContactForm onSubmit={handleSubmit} />    );
  }
}
```

To be clear all this would work with JSX and the only thing that does not work is I cannot pass onSubmit to the exported ContactForm

**Comment by snerks** [comment]
I was initially confused about the mechanics of `handleSubmit`/`onSubmit`...

If I understand it correctly, the method, called `handleSubmit`, is supplied, as a `prop`, by the `reduxForm` Higher-Order Component.

As such, the fact that your custom handler `prop` method is also called `handleSubmit` may be part of the confusion.

I renamed my custom handler `prop` name to something/anything else (`myHandleSave`, for example).

The usage in `ContactForm` then becomes
`const { handleSubmit, myHandleSave } = this.props;`
...
`<form className="form-horizontal" onSubmit={handleSubmit!(myHandleSave)}>`

The usage in the caller is:
`<ContactForm myHandleSave={this.save} />`

I have a (similar) working example here:
[https://github.com/snerks/redux-form-1/blob/master/src/components/ContactForm.tsx](https://github.com/snerks/redux-form-1/blob/master/src/components/ContactForm.tsx)

**Comment by waynebrantley** [comment]
@snerks That of course would work...but there is no need for a custom prop and attaching it.  onSubmit is a property of the form and should be available to ContactForm.  The Redux-Form library properly uses onSubmit event by just doing ``onSubmit={handleSubmit}``

For now, I just used jsx to create my stateless form so that I could use onSubmit and it worked straight away.  The definitions here are just missing that and evidently it is a decent size change.

**Comment by scamden** [comment]
@waynebrantley is right. this is the exact usage pattern used in the getting started docs
![image](https://cloud.githubusercontent.com/assets/1128856/24885932/212ec682-1e06-11e7-8123-201e4e5aa3c2.png)

why would it be hard to add the optional onSubmit prop to the wrapped form component?

**Comment by waynebrantley** [comment]
Until this is fixed, I simply created a JSX file with the form so I can use it as designed.

**Comment by scamden** [comment]
ya that's just a bummer because you lose all type safety, i opted to go with the custom save function as prop route, but obviously typing it consistently with the api would be great

**Comment by corydeppen** [comment]
Is anyone actively working on or considering a fix for this definition/library inconsistency? I'd prefer to use redux-form as designed rather than introduce an unsupported TypeScript-specific hack.

**Comment by virzak** [comment]
Any news on this item?

**Comment by harrisrobin** [comment]
Running into the same issue

**Comment by AleskiWeb** [comment]
I have had success with the following. I think this is what you are trying to achieve?

`LoginScene.tsx`
```typescript
import LoginForm from '../../components/Forms/LoginForm'

class Login extends React.PureComponent<IProps, IState> {
    
    // Submit function in the Parent component
    private customSubmitFunction = (formValues) => {
        console.log('form values are', formValues)
    }

    render () {
        return (
            // ...elements above
            <LoginForm handleSubmit={ this.customSubmitFunction  } />
           //...elements below
        )
    }
}
```

Then in `LoginForm.tsx`
```typescript
interface IProps {
    handleSubmit(value: any): void
}

const loginFormStatic = ({ handleSubmit }: IProps) => (
    <Form>
        <Button onPress={
            (event: any) => {
                handleSubmit(event)
            }
        }>
            <Text>Login</Text>
        </Button>
    </Form>
)

/* tslint:disable-next-line:variable-name */
const LoginForm = reduxForm({
    form: 'login-form',
    validate
  })(loginFormStatic)

export default LoginForm
```

When the form is submitted, pressing the button, then the `customSubmitFunction` is called from the parent. Any help?

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
